mobly/controllers/android_device_lib/snippet_client_v2.py

```diff
@@ -0,0 +1,326 @@
+    match = re.match('^SNIPPET SERVING, PORT ([0-9]+)$', line)
+    if not match:
+      raise errors.ServerStartProtocolError(self._device, line)
+    self.device_port = int(match.group(1))
+
+  def _run_adb_cmd(self, cmd):
+    """Starts a long-running adb subprocess and returns it immediately."""
+    adb_cmd = [adb.ADB]
+    if self._adb.serial:
+      adb_cmd += ['-s', self._adb.serial]
+    adb_cmd += ['shell', cmd]
+    return utils.start_standing_subprocess(adb_cmd, shell=False)
+
+  def _get_persisting_command(self):
+    """Returns the path of a persisting command if available."""
+    for command in [_SETSID_COMMAND, _NOHUP_COMMAND]:
+      try:
+        if command in self._adb.shell(['which', command]).decode('utf-8'):
+          return command
+      except adb.AdbError:
+        continue
+
+    self.log.warning(
+        'No %s and %s commands available to launch instrument '
+        'persistently, tests that depend on UiAutomator and '
+        'at the same time perform USB disconnections may fail.',
+        _SETSID_COMMAND, _NOHUP_COMMAND)
+    return ''
+
+  def _get_user_command_string(self):
+    """Gets the appropriate command argument for specifying device user ID.
+
+    By default, this client operates within the current user. We
+    don't add the `--user {ID}` argument when Android's SDK is below 24,
+    where multi-user support is not well implemented.
+
+    Returns:
+      A string of the command argument section to be formatted into
+      adb commands.
+    """
+    sdk_version = int(self._device.build_info['build_version_sdk'])
+    if sdk_version < 24:
+      return ''
+    return f'--user {self.user_id}'
+
+  def _read_protocol_line(self):
+    """Reads the next line of instrumentation output relevant to snippets.
+
+    This method will skip over lines that don't start with 'SNIPPET ' or
+    'INSTRUMENTATION_RESULT:'.
+
+    Returns:
+      A string for the next line of snippet-related instrumentation output,
+        stripped.
+
+    Raises:
+      errors.ServerStartError: If EOF is reached without any protocol lines
+        being read.
+    """
+    while True:
+      line = self._proc.stdout.readline().decode('utf-8')
+      if not line:
+        raise errors.ServerStartError(
+            self._device, 'Unexpected EOF when waiting for server to start.')
+
+      # readline() uses an empty string to mark EOF, and a single newline
+      # to mark regular empty lines in the output. Don't move the strip()
+      # call above the truthiness check, or this method will start
+      # considering any blank output line to be EOF.
+      line = line.strip()
+      if (line.startswith('INSTRUMENTATION_RESULT:') or
+          line.startswith('SNIPPET ')):
+        self.log.debug('Accepted line from instrumentation output: "%s"', line)
+        return line
+
+      self.log.debug('Discarded line from instrumentation output: "%s"', line)
+
+  def stop(self):
+    """See base class."""
```

Comment:
For the subclass docstrings, we should elaborate a bit instead of pointing to the base class.

e.g. here we can talk about what res we are releasing, which is specific to this impl.

Same for all the other methods